add new guide (#1120)
mattermost/mattermost-developer-documentation: site/content/contribute/expectations/_index.md

---
title: "Section 3 — Contributor expectations"
heading: "Section 3 — Contributor expectations"

> description: ""

Reviewer: Thanks for removing this, @neflyte!

date: "2022-10-01T12:00:00-00:00"
weight: 3
---

## Subsection 1 — Before contributing

1. If you’re working on code or documentation, sign the [Contributor License Agreement](https://mattermost.com/mattermost-contributor-agreement/) so you can be added to the Mattermost [Approved Contributor List](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true). If you’ve included your mailing address in the signed agreement, you might receive a [Limited Edition Mattermost Mug](https://forum.mattermost.com/t/limited-edition-mattermost-mugs/143) as a thank you gift after your first pull request is merged.

> 2. For features that don’t belong as part of a core Mattermost repository:

Reviewer: Good feedback. I agree that a numbered list implies a series of steps that are intended to be followed in their prescribed order. I've provided content suggestions that convert these numbered steps to sub-steps as "a, b, c, d". A future iteration could include revisiting this section to remove the ordered approach altogether by changing these contribution areas to bullet points or paragraphs instead.

> 2. For features that don’t belong as part of a core Mattermost repository:

Reviewer: @jordanafung - I've noted that we should revisit this section in a future revision.

    1. You could [create an app]({{< ref "/integrate/apps/" >}}), which connects to Mattermost’s functionality while being hosted elsewhere.
    2. External applications that don’t rely on customizing the Mattermost user experience can reach the data they need through [incoming]({{< ref "/integrate/webhooks/incoming" >}}) and [outgoing]({{< ref "/integrate/webhooks/outgoing" >}}) webhooks or [the API](https://api.mattermost.com/).
    3. External functionality activated from within Mattermost might work best as a [slash command]({{< ref "/integrate/slash-commands/" >}}).

>     4. [Plugins]({{< ref "/integrate/plugins/" >}}) enable you to modify the fabric of the Mattermost UI/UX completely but come with the highest level of development overhead and must be written in Go and React.

Reviewer: ```suggestion
    d. You can modify the fabric of the Mattermost UI/UX completely using [plugins]({{< ref "/integrate/plugins/" >}}). However please note that plugin development comes with the highest level of overhead and must be written in Go and React.
```

Reviewer 2: "modify the fabric" sounds a bit vague and magically lol
Alternate suggestion:
You can extend, modify and deeply integrate with the Mattermost UI/UX by using

Reviewer 2: ```suggestion
    4. You can extend, modify, and deeply integrate with the Mattermost server, its apps and its UI/UX by using [plugins]({{< ref "/integrate/plugins/" >}}). However please note that plugin development comes with the highest level of overhead and must be written in Go and React.
```

>     5. If you need to take Mattermost with you into another app, perhaps you’re looking for [the embed guide]({{< ref "/integrate/customization/embedding/" >}}).

Reviewer: Given that the preceding 4 steps are specific ways to contribute to areas of Mattermost beyond core repos, this line feels more like a tip or note rather than a fifth contribution opportunity.

Reviewer 2: Please see my suggestion below; this may resolve it

Reviewer 2: ```suggestion
    5. If you need to take Mattermost with you into another app, perhaps you’re looking for [the embed guide]({{< ref "/integrate/customization/embedding/" >}}).
```

>     5. If you need to take Mattermost with you into another app, perhaps you’re looking for [the embed guide]({{< ref "/integrate/customization/embedding/" >}}).

Reviewer 2: ```suggestion
    5. You can use Mattermost from other applications, by [embedding]({{< ref "/integrate/customization/embedding/" >}}) and launching Mattermost within other applications and mobile apps.
```

> 3. Read the Markdown files in the root of the repo you’re working in to learn about processes unique to that repo.

Reviewer: ```suggestion
3. Identify which repository you need to be working in, then review the README located within the root of the repository to learn more about getting started with your contribution.
```

> 4. Figure out which repo you need to be working in. With content, you should be able to click the Edit in GitHub button on the published page to find the right repo, but with code it can be somewhat more difficult. You might even have to contribute to multiple repos, likely starting in the server repo. Here are the core repos:

Reviewer: Let's add part of this step about contributing content updates to our documentation suite as a fifth way to contribute to Mattermost outside of core repositories.

Then, let's move the list of core repositories to the step above about identifying the correct repo OR incorporate this list of core repos to introduction where we introduce that there's more than just core contributions to focus on.

Reviewer 2: I've given an alternate in the suggestion below https://github.com/mattermost/mattermost-developer-documentation/pull/1120/files#r984157530

>     - [Server]({{< ref "/contribute/more-info/server/" >}}) - Highly-scalable Mattermost installation written in Go

Reviewer: ```suggestion
    - [Server]({{< ref "/contribute/more-info/server/" >}}): Highly-scalable Mattermost installation written in Go.
```

> 4. Figure out which repo you need to be working in. With content, you should be able to click the Edit in GitHub button on the published page to find the right repo, but with code it can be somewhat more difficult. You might even have to contribute to multiple repos, likely starting in the server repo. Here are the core repos:

Reviewer 2: ```suggestion
4. To contribute to documentation, you should be able to get to its repo by clicking the "Edit in Github" button in the top right of its respective published page.
5. Otherwise you can contribute to these Mattermost Core repos:
```

Reviewer 2: Actually thinking we can integrate point 5 into point 2?

>     - [Web App]({{< ref "/contribute/more-info/webapp/" >}}) - JavaScript client app built on React and Redux

Reviewer: ```suggestion
    - [Web App]({{< ref "/contribute/more-info/webapp/" >}}): JavaScript client app built on React and Redux.
```

Reviewer: Built on React Native?

>     - [Mobile Apps]({{< ref "/contribute/more-info/mobile/" >}}) - JavaScript client apps for Android and iOS built on React Native

Reviewer: ```suggestion
    - [Mobile Apps]({{< ref "/contribute/more-info/mobile/" >}}): JavaScript client apps for Android and iOS built on React Native.
```

>     - [Desktop App]({{< ref "/contribute/more-info/desktop/" >}}) - An Electron wrapper around the web app project that runs on Windows, Linux, and macOS

Reviewer: ```suggestion
    - [Desktop App]({{< ref "/contribute/more-info/desktop/" >}}): An Electron wrapper around the web app project that runs on Windows, Linux, and macOS.
```

>     - [Core Plugins]({{< ref "/contribute/more-info/plugins/" >}}) - A core set of officially-maintained plugins that provide a variety of improvements to Mattermost.

Reviewer: ```suggestion
    - [Core Plugins]({{< ref "/contribute/more-info/plugins/" >}}). A core set of officially-maintained plugins that provide a variety of improvements to Mattermost.
```

>     - Core Integrations - Major Mattermost features including [Focalboard]({{< ref "/contribute/more-info/focalboard/" >}}) and [Playbooks](https://github.com/mattermost/mattermost-plugin-playbooks).

Reviewer: ```suggestion
    - Core Integrations: Mattermost product verticals including [Boards]({{< ref "/contribute/more-info/focalboard/" >}}) and [Playbooks](https://github.com/mattermost/mattermost-plugin-playbooks).
```

> ## Subsection 2 — During the contribution process

Reviewer: ```suggestion
## During the contribution process
```

> 1. Check in regularly with your PR to review and respond to feedback. 

Reviewer 2: ```suggestion
1. Check in regularly with your Pull Request (PR) to review and respond to feedback. 
```

2. Thoroughly document what you’re doing. This way, future contributors will be easily able to pick up on your work (including yourself). This is especially helpful if you need to step back from a PR — in that case, just tell your supervisor how the next contributor can find your documentation.

> 3. PRs should represent projects, both in code and in content. Keep unrelated tasks to different PRs.

Reviewer 2: ```suggestion
3. Each PR should represent a single project, both in code and in content. Keep unrelated tasks to different PRs.
```

4. Make your PR titles and commit messages descriptive! If you make reviewer’s lives easier by briefly describing the project in the PR title and in your commit messages, you’ll likely get faster responses, less clarifying questions, and better advice.

## Subsection 3 — While writing code

> 1. Always thoroughly test your contributions! Mattermost is a complex web of programs, so even insignificant changes have the potential to break big features. Here’s our guide on end-to-end testing.

Reviewer 2: ```suggestion
1. Always thoroughly test your contributions! Please see our [testing guidelines]{{< ref "/more-info/getting-started/test-guideline/" >}}). We recommend the following best practices for your contribution:
```

## Subsection 4 — While writing content

1. Respect the time and intelligence of your audience.
    1. Be clear and concise without being blunt or curt.
    2. Prefer proper or common nouns first, [then gender-neutral pronouns](https://apastyle.apa.org/style-grammar-guidelines/grammar/singular-they).
    3. Give each page a general summary at the top, right under the first heading.
    4. Treat every page like page one. It’s more likely that your audience found the content through a Google search than by flipping through the content repository where you’re writing.
    5. Prefer short, single-point paragraphs. If they’re small enough once condensed and logically fit together, it might make sense to turn them into a bulleted list. Readers are going to skim your content, so organize your presentation to accommodate them.
    6. Use technical terms where it’s warranted, as long as its still simple to understand for those who don’t already know the jargon.

>     7. Avoid cliches, especially those that don’t mean anything significant.

Reviewer: Nice catch!

    8. Don’t make assumptions about what readers find “easy” or “simple”.
    9. Let the product speak for itself — avoid directly praising Mattermost.
    10. Write for an international audience.
2. Use tone, word choice, and grammar appropriate for the situation.
    1. Use the active voice.
    2. Negative sentences work best when used sparingly.
    3. Prefer the second-person (”you”) to the first (”I” and “we”).
    4. Use words for “zero” through “nine” and digits for 10 and up. [There are plenty of exceptions to this, though.](https://apastyle.apa.org/style-grammar-guidelines/numbers/numerals)
    5. Use the Oxford comma (that’s the last comma before the conjunction in a list of at least three things).
    6. Use sentence case everywhere.
    7. Present the company, the staff, the community, and the product as friendly for developers, IT folks, and operations specialists.

>     8. American English is the standard here.

Reviewer: See note above. We'll move these guidelines from this PR to the Mattermost Handbook.
